File: ruby/lib/simple_cbow.rb
require "numo/narray"
require "mat_mul"
require "softmax_with_loss"
require "embedding"

class SimpleCBow
  attr_reader :params, :grads, :word_vecs

  def initialize(vocab_size, hidden_size)
    v, h = vocab_size, hidden_size

    # 重みの初期化
    w_in = 0.01 * Numo::DFloat.new(v, h).rand
    w_out = 0.01 * Numo::DFloat.new(h, v).rand

    # レイヤの生成
    @in_layer0 = Embedding.new(w_in)
    @in_layer1 = Embedding.new(w_in)
    # @in_layer0 = MatMul.new(w_in)
    # @in_layer1 = MatMul.new(w_in)
    @out_layer = MatMul.new(w_out)
    @loss_layer = SoftmaxWithLoss.new

    # すべての重みと勾配をリストにまとめる
    layers = [@in_layer0, @in_layer1, @out_layer]
    @params, @grads = layers.reduce([[], []]) do |acc, layer|
      [acc[0] + layer.params, acc[1] + layer.grads]
    end

    # メンバ変数に単語の分散表現を設定
    @word_vecs = w_in
  end

  def forward(contexts, target)
    h0 = @in_layer0.forward(contexts[true, 0])
    h1 = @in_layer1.forward(contexts[true, 1])
    h = (h0 + h1) * 0.5
    score = @out_layer.forward(h)
    loss = @loss_layer.forward(score, target)
    return loss
  end

  def backward(dout=1)
    ds = @loss_layer.backward(dout)
    da = @out_layer.backward(ds)
    da *= 0.5
    @in_layer1.backward(da)
    @in_layer0.backward(da)
    nil
  end
end

File: gpu/ch03/simple_cbow.py
import cupy as cp

from ch01.layers import MatMul, SoftmaxWithLoss

class SimpleCBOW:
    def __init__(self, vocab_size, hidden_size):
        V, H = vocab_size, hidden_size

        #重みの初期化
        W_in = 0.01 * cp.random.randn(V, H).astype(cp.float32)
        W_out = 0.01 * cp.random.randn(H, V).astype(cp.float32)

        #レイヤの生成
        self.in_layer0 = MatMul(W_in)
        self.in_layer1 = MatMul(W_in)
        self.out_layer = MatMul(W_out)
        self.loss_layer = SoftmaxWithLoss()

        # すべての重みと購買をリストにまとめる
        layers = [self.in_layer0, self.in_layer1, self.out_layer]
        self.params, self.grads = [],[]
        for layer in layers:
            self.params += layer.params
            self.grads += layer.grads

        #メンバ変数に単語の分散表現を追加
        self.word_vecs = W_in

    def forward(self, contexts, target):
        h0 = self.in_layer0.forward(contexts[:, 0])
        h1 = self.in_layer1.forward(contexts[:, 1])
        h = (h0 + h1) * 0.5
        score = self.out_layer.forward(h)
        loss = self.loss_layer.forward(score, target)
        return loss

    def backward(self, dout=1):
        ds = self.loss_layer.backward(dout)
        da = self.out_layer.backward(ds)
        da *= 0.5
        self.in_layer1.backward(da)
        self.in_layer0.backward(da)
        return None
